src/models.ts

```diff
@@ -133,6 +138,11 @@ function wrapResponseError<E extends Error>(e: E): void {
 		throw new errors.UnauthorisedError(e);
 	case 404:
 		throw new errors.ObjectDoesNotExistError(e);
+	case 409:
+		if (typeof error.message === 'string' && /unique/i.test(error.message)) {
```

Comment:
I'm confused, so we've fixed the response code in the API in order to not have clients parse the response and we made it such that parsing is still required? Why doesn't the API just return a different response code for each case, given how common they are?

Reply:
I actually based this on [this pinejs-client PR comment](https://github.com/balena-io/pinejs-client-js/pull/81#discussion_r301708160). So I also want to hear @Page- 's thoughts on this.

Reply:
We've improved the status code to use the correct one for a unique constraint error @dfunckt, but there's other types of error that should also use a 409 (eg a foreign key constraint error) so to be sure it's a unique constraint error we need to check the extra info - that's only necessary if we care about being sure it's a unique error though

Reply:
@Page- I get that, but practically what's the difference? We changed the status code but kept the fragility.

Reply:
The new status code reduces the ambiguity, both in terms of the 409 making the class of error clearer for this case but also for the 404 becoming less overused.  But yes there are only so many http status codes so each one is not unique to a specific error just to a specific class, and for this case I think we care about the specific error on top of just the class,

As for the fragility, knowing that it's a 409 + unique is less fragile than 404 + unique which can have more false positives (and compared to just checking 404 vs just 409 then it's a massive improvement).  In the end though the real solution is that this should just be a single request rather than having client-side logic but for now this is at least a better compromise

Reply:
I still don't see the improvement in this specific case -- the only case one will get a 404 during this POST is for all the same reasons they'd get a 409. I can see the bigger picture and I agree it's vastly better, but I still can't justify all this fuss when the new solution is practically the same as before. As you say, a single request is what is actually needed, and I don't see the point changing this until we get one.